src/wallet/scriptpubkeyman.cpp

```diff
@@ -0,0 +1,1448 @@
+    case IsMineResult::SPENDABLE:
+        return ISMINE_SPENDABLE;
+    }
+    assert(false);
+}
+
+bool LegacyScriptPubKeyMan::HasEncryptedData() const
+{
+    LOCK(cs_KeyStore);
+    return HasEncryptionKeys() || !mapCryptedKeys.empty();
+}
+
+bool LegacyScriptPubKeyMan::CheckDecryptionKey(const CKeyingMaterial& master_key, bool accept_no_keys)
+{
+    {
+        LOCK(cs_KeyStore);
+        if (!mapKeys.empty())
+            return false;
+
+        bool keyPass = mapCryptedKeys.empty(); // Always pass when there are no encrypted keys
+        bool keyFail = false;
+        CryptedKeyMap::const_iterator mi = mapCryptedKeys.begin();
+        for (; mi != mapCryptedKeys.end(); ++mi)
+        {
+            const CPubKey &vchPubKey = (*mi).second.first;
+            const std::vector<unsigned char> &vchCryptedSecret = (*mi).second.second;
+            CKey key;
+            if (!DecryptKey(master_key, vchCryptedSecret, vchPubKey, key))
+            {
+                keyFail = true;
+                break;
+            }
+            keyPass = true;
+            if (fDecryptionThoroughlyChecked)
+                break;
+        }
+        if (keyPass && keyFail)
+        {
+            LogPrintf("The wallet is probably corrupted: Some keys decrypt but not all.\n");
+            throw std::runtime_error("Error unlocking wallet: some keys decrypt but not all. Your wallet file may be corrupt.");
+        }
+        if (keyFail || (!keyPass && !accept_no_keys))
+            return false;
+        fDecryptionThoroughlyChecked = true;
+    }
+    return true;
+}
+
+bool LegacyScriptPubKeyMan::Encrypt(const CKeyingMaterial& master_key, WalletBatch* batch)
+{
+    LOCK(cs_KeyStore);
+    encrypted_batch = batch;
+    if (!mapCryptedKeys.empty()) {
+        encrypted_batch = nullptr;
+        return false;
+    }
+
+    KeyMap keys_to_encrypt;
+    keys_to_encrypt.swap(mapKeys); // Clear mapKeys so AddCryptedKeyInner will succeed.
+    for (const KeyMap::value_type& mKey : keys_to_encrypt)
+    {
+        const CKey &key = mKey.second;
+        CPubKey vchPubKey = key.GetPubKey();
+        CKeyingMaterial vchSecret(key.begin(), key.end());
+        std::vector<unsigned char> vchCryptedSecret;
+        if (!EncryptSecret(master_key, vchSecret, vchPubKey.GetHash(), vchCryptedSecret)) {
+            encrypted_batch = nullptr;
+            return false;
+        }
+        if (!AddCryptedKey(vchPubKey, vchCryptedSecret)) {
+            encrypted_batch = nullptr;
+            return false;
+        }
+    }
+    mapKeys.clear();
+    encrypted_batch = nullptr;
+    return true;
+}
+
+bool LegacyScriptPubKeyMan::GetReservedDestination(const OutputType type, bool internal, CTxDestination& address, int64_t& index, CKeyPool& keypool)
+{
+    if (!CanGetAddresses(internal)) {
+        return false;
+    }
+
+    {
+        LOCK(cs_KeyStore);
+        if (!ReserveKeyFromKeyPool(index, keypool, internal) && !IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
```

Comment:
Why is `&& !IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)` added here? This seems broken because why would you want to return true from GetReserved if reserving a key failed? But it also seems like it should not be able to make any difference because the CanGetAddresses call above would return false if private keys were disabled.

I'd suggest either dropping this condition adding an explanatory comment. Tests do seem to pass if this is dropped.

Also, I'm not sure if the CanGetAddresses check added above is more than just a sanity check, but if it is actually an expected condition, it'd be useful to have a comment saying when it's expected to trigger. It seems like a fine check, but it's not clear what the motivation was for adding it. Tests do seem to pass without it.

Reply:
Removed. I don't remember why they were added, could have been due to some previous change that was since changed so they are no longer needed.

I believe the CanGetAddresses check is for sanity.